src/ingestion/ec2_describe_instances.py

```diff
@@ -0,0 +1,128 @@
+    try:
+        dest_role = global_aws_client.assume_role(
+            RoleArn=target_role,
+            RoleSessionName=os.environ['EC2_INSTANCE_LIST_AWS_AUDIT_DESTINATION_ROLE_SESSION_NAME'],
+            ExternalId=os.environ['AWS_AUDIT_DESTINATION_ROLE_ARN_EXTERNALID']
+        )
+    except Exception as e:
+        return None
+    ec2_session = boto3.Session(
+        aws_access_key_id=dest_role['Credentials']['AccessKeyId'],
+        aws_secret_access_key=dest_role['Credentials']['SecretAccessKey'],
+        aws_session_token=dest_role['Credentials']['SessionToken']
+    )
+    return ec2_session
+
+def get_accounts_list(sf_client):
+    accounts_list = []
+    query = f"select account:Id::number from {AWS_ACCOUNTS_TABLE} where timestamp = (select max(timestamp) from {AWS_ACCOUNTS_TABLE});"
+    res = sf_client.cursor().execute(query).fetchall()
+    for account in res:
+        accounts_list.append(account[0])
+    return accounts_list
+
+
+LOAD_INSTANCE_LIST_QUERY = f"""
+INSERT INTO {INSTANCES_TABLE} (timestamp, instance)
+SELECT '{{snapshotclock}}'::timestamp_ltz, PARSE_JSON(column1)
+FROM VALUES {{format_string}}
+"""
+
+def get_data_worker(account):
+    ec2_session = get_aws_client(account)
+    if ec2_session:
+        instances = []
+        ec2_regions = [region['RegionName'] for region in boto3.client('ec2').describe_regions()['Regions']]
+        for region in ec2_regions:
+            ec2_client = ec2_session.client('ec2', region_name=region)
+            paginator = ec2_client.get_paginator('describe_instances')
+            page_iterator = paginator.paginate()
+            for page in page_iterator:
+                for instance_array in page['Reservations']:
+                    instances.extend(instance_array['Instances'])
+        instance_list = [json.dumps({**instance,"AccountId":account}, default=str) for instance in instances]
+        if len(instances):
+            sf_client = get_snowflake_client()
+            query = LOAD_INSTANCE_LIST_QUERY.format(
+                snapshotclock=datetime.datetime.utcnow().isoformat(),
+                format_string=", ".join(["(%s)"] * len(instance_list)))
+            sf_client.cursor().execute(query, instance_list)
+
+def get_data(accounts_list):
+    start = datetime.datetime.now()
+    with Pool(processes=4) as pool:
+        pool.map(get_data_worker, accounts_list)
+    end = datetime.datetime.now()
+    print(f"start: {start} end: {end} total: {(end - start).total_seconds()}")
+
+def main():
+    sf_client = get_snowflake_client()
+    current_time = datetime.datetime.now(datetime.timezone.utc)
+    last_time = sf_client.cursor().execute(f'SELECT max(timestamp) FROM {INSTANCES_TABLE}').fetchall()[0][0]
+    if last_time is None or (current_time - last_time).total_seconds() > 86400:
+        accounts_list = get_accounts_list(sf_client)
+        get_data(accounts_list)
+    else:
+        print("It's not time yet!")
+
+global_aws_client=get_global_aws_client()
```

Comment:
so unless there's a reason otherwise, you don't want python modules to do anything when you import them, and everything to happen after you call main(). here, you can do —

    if __name__ == '__main__':
        main()

and put `global_aws_client=get_global_aws_client()` into `get_aws_client`, since that's the only place it's used